File: 15-3sum/3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        HashMap<Integer,Integer>map=new HashMap<>();
        List<List<Integer>>ans=new ArrayList<>();
        int n=nums.length;
        for(int i=0;i<n;i++){
            map.put(nums[i],i);
        }
        HashSet<List<Integer>> set = new HashSet<>();
        for(int i=0;i<n-1;i++){
            for(int j=i+1;j<n;j++){
                int sum=-1*(nums[i]+nums[j]);
                if(map.containsKey(sum) && i!=map.get(sum) && j!=map.get(sum)){
                    List<Integer>a=new ArrayList<>();
                    a.add(nums[i]);
                    a.add(nums[j]);
                    a.add(sum);
                    Collections.sort(a);
                    if(!set.contains(a)) {
                        ans.add(new ArrayList<>(a));
                        set.add(a);
                    }
                }
                // map.put(nums[i],i);
            }
        }
        return ans;
    }
}

File: 15-3sum/3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> ans;
        
        sort(nums.begin(),nums.end());
        for(int i=0;i<nums.size()-2;i++){
            if(i>0 && nums[i]==nums[i-1]) continue;
            int target=0-nums[i];
            int low=i+1;
            int high=nums.size()-1;
            while(low<high){
                if(nums[low]+nums[high]==target){
                    ans.push_back({nums[i],nums[low],nums[high]});
                    low++;
                    high--;
                    while(low<nums.size() && nums[low]==nums[low-1])low++;
                    while(high<nums.size()-1 && nums[high]==nums[high+1])high--;
                }
                else if(nums[low]+nums[high]<target){
                    low++;
                }
                else{
                    high--;
                }
            }
        }
        return ans;
    }
};